File: legacy/src/run-turn.ts
/**
 * OpenCode turn execution: prompt, waitIdle, stall recovery, session rotate.
 * Uses SDK status/events/summarize/noReply — no process-name heuristics.
 */
import fs from "node:fs"
import { PROVIDER_ID, bareModel } from "./config.ts"
import type { Api, EventBus } from "./opencode.ts"
import { probeSession, type SessionProbeMeta } from "./session-probe.ts"
import type { AgentRef } from "./run-types.ts"
import { sleep } from "./run-types.ts"
import { trace } from "./trace.ts"

export type TurnDeps = {
  api: Api
  bus: EventBus
  stallMs: number
  runId: string
  workerSessionFile: string
  systemSessionFile?: string
  isStopping: () => boolean
  markActivity: () => void
  lastActivityAt: () => number
  log: (msg: string) => void
  onSessionRotated?: (agent: AgentRef) => void
  /** Probe+archive worker session before id changes (stall/size rotate). */
  archiveWorkerBeforeRotate?: (agent: AgentRef) => void | Promise<void>
  /** Optional last summary text to inject into a fresh session after rotate. */
  lastRotateSummary?: { get: () => string; set: (s: string) => void }
  /**
   * When true, external Aborted is terminal (no re-prompt).
   * Used when system watch deliberately aborts a stuck worker.
   */
  suppressExternalAbortRetry?: () => boolean
}

function isContextSizeError(msg: string): boolean {
  return /bad request|context.?overflow|context.?length|too large|token|413\b|payload/i.test(msg)
}

/**
 * OpenCode session aborted by human (TUI Esc), concurrent prompt on same session,
 * or external cancel — not a size/stall fault. Prefer re-prompt same session once.
 */
function isExternalAbortError(msg: string): boolean {
  return /\babort(ed|ing)?\b|\bcancell?ed\b|\binterrupted\b/i.test(msg) && !/stall:/i.test(msg)
}

export async function lastAssistantText(api: Api, agent: AgentRef): Promise<string> {
  try {
    const messages = await api.sessionMessages(agent.directory, agent.sessionID)
    const last = [...messages].reverse().find((m: any) => m?.info?.role === "assistant")
    return (last?.parts ?? [])
      .filter((p: any) => p?.type === "text" && p.text)
      .map((p: any) => String(p.text))
      .join("\n")
      .trim()
  } catch {
    return ""
  }
}

async function waitUntilNotBusy(
  api: Api,
  directory: string,
  sessionID: string,
  isStopping: () => boolean,
): Promise<void> {
  for (let i = 0; i < 40; i++) {
    if (isStopping()) return
    try {
      const statuses = await api.sessionStatus(directory)
      const st = statuses[sessionID]
      if (!st || st.type === "idle") return
    } catch {
      return
    }
    await sleep(500)
  }
}

export async function rotateSession(deps: TurnDeps, agent: AgentRef): Promise<void> {
  if (agent.role === "worker" && deps.archiveWorkerBeforeRotate) {
    try {
      await deps.archiveWorkerBeforeRotate(agent)
    } catch (err) { trace("runTurn.rotateSession", err) }
  }

  // Primary path: SDK session.fork — the native "continue from here" mechanism.
  // The forked session inherits a compacted view of the parent's context,
  // giving a fresh context window with the essential summary. More reliable
  // than our manual summarize → create → inject pattern.
  try {
    const forkedId = await deps.api.sessionFork(agent.directory, agent.sessionID)
    agent.sessionID = forkedId
    deps.onSessionRotated?.(agent)
    deps.log(`  [host] session.fork ok for ${agent.role} — new session ${forkedId.slice(0, 16)}…`)
    return
  } catch (forkErr) {
    deps.log(
      `  [host] session.fork failed: ${forkErr instanceof Error ? forkErr.message : String(forkErr)}`.slice(0, 160),
    )
  }

  // Fallback: SDK summarize → create → inject (the old manual pattern).
  // Used when fork is unavailable or fails. If summarize also fails (session
  // too big), create a fresh session with a minimal host-written continuity note.
  let summaryNote = ""
  let summarizeOk = false
  try {
    const ok = await deps.api.sessionSummarize(agent.directory, agent.sessionID, {
      providerID: PROVIDER_ID,
      modelID: bareModel(agent.model),
    })
    if (ok) {
      summarizeOk = true
      summaryNote = await lastAssistantText(deps.api, agent)
      if (summaryNote) {
        deps.lastRotateSummary?.set(summaryNote.slice(0, 6_000))
        deps.log(`  [host] session.summarize ok for ${agent.role} (${summaryNote.length} chars)`)
      } else {
        deps.log(`  [host] session.summarize accepted for ${agent.role}`)
      }
    }
  } catch (err) {
    deps.log(
      `  [host] session.summarize failed (session may be too big): ${err instanceof Error ? err.message : String(err)}`.slice(0, 160),
    )
  }

  const prevSummary = deps.lastRotateSummary?.get() || summaryNote
  const session = await deps.api.createSession(agent.directory, `swarm ${deps.runId} ${agent.role} (rotated)`)
  agent.sessionID = session.id
  deps.onSessionRotated?.(agent)

  if (prevSummary.trim() && summarizeOk) {
    try {
      await deps.api.sessionInjectContext(
        agent.directory,
        agent.sessionID,
        [
          `[host] Prior ${agent.role} session was rotated. Continuity summary from OpenCode:`,
          prevSummary.slice(0, 6_000),
        ].join("\n\n"),
        { providerID: PROVIDER_ID, modelID: bareModel(agent.model) },
      )
      deps.log(`  [host] injected rotate summary into new ${agent.role} session (noReply)`)
    } catch (err) {
      deps.log(
        `  [host] noReply inject failed: ${err instanceof Error ? err.message : String(err)}`.slice(0, 160),
      )
    }
  } else {
    try {
      await deps.api.sessionInjectContext(
        agent.directory,
        agent.sessionID,
        [
          `[host] Prior ${agent.role} session was rotated (summarize failed — session was too large).`,
          `You are continuing an autonomous run. Read MISSION.md, DIALOGUE.md, and MEMORY.md`,
          `in the run folder for full context. The conversation history is on disk.`,
        ].join("\n"),
        { providerID: PROVIDER_ID, modelID: bareModel(agent.model) },
      )
      deps.log(`  [host] injected fallback continuity note into new ${agent.role} session (noReply)`)
    } catch (err) {
      deps.log(
        `  [host] fallback inject failed: ${err instanceof Error ? err.message : String(err)}`.slice(0, 160),
      )
    }
  }

  deps.log(`  [host] rotated session for ${agent.role} (fresh context, fallback path)`)
}

/**
 * Prompt + wait with Bad Request / stall recovery.
 * No wall-clock kill of healthy long tools — only zero bus activity for stallMs.
 * Optional `system` is sticky identity (OpenCode system field) — use for lead role.
 */
export async function runTurn(
  deps: TurnDeps,
  agent: AgentRef,
  prompt: string,
  opts?: { system?: string },
): Promise<{ text: string; secs: number }> {
  const maxAttempts = 3
  let lastErr: Error | undefined

  for (let attempt = 1; attempt <= maxAttempts; attempt++) {
    const t0 = Date.now()
    try {
      deps.markActivity()
      const idle = deps.bus.waitIdle(agent.directory, agent.sessionID, () => deps.isStopping())
      await deps.api.promptAsync(agent.directory, agent.sessionID, {
        model: { providerID: PROVIDER_ID, modelID: bareModel(agent.model) },
        ...(opts?.system ? { system: opts.system } : {}),
        parts: [{ type: "text", text: prompt }],
      })

      let finished = false
      const work = idle.then(() => {
        finished = true
      })
      const stallWatch = (async () => {
        while (!finished) {
          await sleep(15_000)
          if (finished || deps.isStopping()) return
          // Prefer last *session* bus event for quiet; do not let global markActivity hide silence.
          const lastBus = deps.bus.lastActivityFor(agent.sessionID) || deps.lastActivityAt()
          const quietMs = Date.now() - lastBus
          // Stuck tool accounting: no events for half stall window → drop "running" flag.
          if (deps.bus.clearStaleRunningTools(agent.sessionID, Math.min(deps.stallMs / 2, 10 * 60_000))) {
            deps.log(
              `  [host] cleared stale running-tool flag on ${agent.role} (no bus events ≥${Math.round(Math.min(deps.stallMs / 2, 10 * 60_000) / 60_000)}m)`,
            )
          }
          // Fresh running tools with recent events = healthy.
          if (deps.bus.hasRunningTools(agent.sessionID) && quietMs < 2 * 60_000) {
            deps.markActivity()
            continue
          }
          // Quiet too long: stall even if status still busy (stuck generation / hung bash).
          if (quietMs >= deps.stallMs) {
            let detail = "no bus events"
            try {
              const act = await deps.api.sessionIsActive(agent.directory, agent.sessionID)
              detail = act.active ? `status still ${act.detail}` : act.detail
            } catch {}
            throw new Error(
              `stall: no OpenCode bus events for ${Math.round(quietMs / 60_000)}m on ${agent.role} (${detail}; threshold ${Math.round(deps.stallMs / 60_000)}m)`,
            )
          }
          // Still within window and session busy with some recent event — keep waiting.
          try {
            const act = await deps.api.sessionIsActive(agent.directory, agent.sessionID)
            if (act.active) {
              // do not markActivity — that would reset quiet and block stall forever
              continue
            }
          } catch (err) { trace("runTurn.stallWatch", err) }
        }
      })()
      await Promise.race([work, stallWatch])
      if (!finished) await work

      const text = await lastAssistantText(deps.api, agent)
      const secs = Math.round((Date.now() - t0) / 1000)
      const oneLine = text.replace(/\s+/g, " ").trim()
      if (oneLine) deps.log(`  [reply:${agent.role}] ${oneLine.slice(0, 300)}`)
      deps.log(`  [metric] ${agent.role} turn ${secs}s`)
      return { text, secs }
    } catch (err) {
      lastErr = err instanceof Error ? err : new Error(String(err))
      const msg = lastErr.message
      if (deps.isStopping() || /stopped/i.test(msg)) throw lastErr

      deps.log(`  [host] turn error attempt ${attempt}/${maxAttempts}: ${msg.slice(0, 300)}`)

      // External abort: session already cancelled — do not thrash with another abort/rotate.
      // Re-prompt the same session so mid-turn work can continue after human TUI interference.
      // Exception: deliberate watch abort of a stuck worker — do not re-prompt (that undoes STOP).
      if (isExternalAbortError(msg)) {
        if (deps.suppressExternalAbortRetry?.()) {
          deps.log(
            `  [host] external abort on ${agent.role} (watch/lead abort) — terminal, no re-prompt`,
          )
          throw lastErr
        }
        deps.log(
          `  [host] external abort on ${agent.role} (human interrupt / concurrent session use) — wait idle, re-prompt same session (no rotate)`,
        )
        try {
          await waitUntilNotBusy(deps.api, agent.directory, agent.sessionID, deps.isStopping)
        } catch (err) { trace("runTurn.waitUntilNotBusy", err) }
        if (attempt < maxAttempts) {
          await sleep(1_500 * attempt)
          continue
        }
        break
      }

      await deps.api.abort(agent.directory, agent.sessionID)
      try {
        await waitUntilNotBusy(deps.api, agent.directory, agent.sessionID, deps.isStopping)
      } catch (err) { trace("runTurn.abortWait", err) }

      // Stall: first recovery re-prompts same session (keep context); later stalls rotate.
      if (/stall:/i.test(msg)) {
        if (attempt === 1) {
          deps.log(
            `  [host] stall soft-recover on ${agent.role} — re-prompt same session (no rotate yet)`,
          )
          await sleep(2_000)
          continue
        }
        deps.log(`  [host] rotating session after repeated stall`)
        await rotateSession(deps, agent)
        await sleep(1_000)
        continue
      }

      if (isContextSizeError(msg)) {
        deps.log(`  [host] rotating session after size/Bad Request`)
        await rotateSession(deps, agent)
        await sleep(1_000)
        continue
      }

      // Other errors: rotate only on the last retry, not every failure.
      if (attempt === maxAttempts - 1) {
        await rotateSession(deps, agent)
      } else if (attempt < maxAttempts) {
        await sleep(2_000 * attempt)
      }
    }
  }
  throw lastErr ?? new Error("turn failed")
}

/** Rotate worker when probe is near this many messages (session dump becomes useless). */
const WORKER_ROTATE_MSG_THRESHOLD = 120

export async function captureWorkerSession(
  deps: TurnDeps,
  worker: AgentRef,
): Promise<SessionProbeMeta> {
  const { meta } = await probeSession(deps.api.client, {
    role: "worker",
    sessionID: worker.sessionID,
    directory: worker.directory,
    dumpPath: deps.workerSessionFile,
    // Recent window for the lead; full history lives in sessions/ archives when useful.
    maxChars: 150_000,
    messageLimit: 80,
    runId: deps.runId,
    redact: true,
  })
  deps.log(
    `  [host:session] worker probe: messages=${meta.messageCount} tools=${meta.toolCalls} errors=${meta.toolErrors} status=${meta.status} → ${meta.dumpPath}` +
      (meta.error ? ` (${meta.error.slice(0, 120)})` : ""),
  )
  return meta
}

/** Probe system/lead session for postmortem archives (not required for worker handoff). */
export async function captureSystemSession(
  deps: TurnDeps,
  system: AgentRef,
): Promise<SessionProbeMeta | null> {
  const dumpPath = deps.systemSessionFile
  if (!dumpPath) return null
  const { meta } = await probeSession(deps.api.client, {
    role: "system",
    sessionID: system.sessionID,
    directory: system.directory,
    dumpPath,
    maxChars: 120_000,
    messageLimit: 60,
    runId: deps.runId,
    redact: true,
  })
  deps.log(
    `  [host:session] system probe: messages=${meta.messageCount} tools=${meta.toolCalls} errors=${meta.toolErrors} status=${meta.status} → ${meta.dumpPath}` +
      (meta.error ? ` (${meta.error.slice(0, 120)})` : ""),
  )
  return meta
}

/**
 * Whether host should rotate the worker session for a fresh episode.
 * session.fork inherits the parent's message count, so we check growth
 * since the last rotation baseline instead of the absolute count.
 */
export function shouldRotateWorker(
  meta: SessionProbeMeta | null,
  emptyShip: boolean,
  emptyStreak: number,
  rotateMsgBase?: number,
): boolean {
  if (emptyShip && emptyStreak >= 1) return true
  if (!meta) return false
  const base = rotateMsgBase ?? 0
  const growth = meta.messageCount - base
  if (growth >= WORKER_ROTATE_MSG_THRESHOLD) return true
  return false
}



/**
 * True when we already dumped this worker session after the last ship.
 * Skip re-probe at next cycle start — the lead still has WORKER_SESSION.md on disk
 * for a deep review; host just avoids rewriting the same dump.
 * Re-probe after session rotate or if the dump file is missing/empty.
 */
export function isWorkerProbeFresh(
  worker: AgentRef,
  last: SessionProbeMeta | null,
  dumpPath: string,
): boolean {
  if (!last) return false
  if (last.sessionID !== worker.sessionID) return false
  if (last.directory !== worker.directory) return false
  if (last.error) return false
  try {
    if (!fs.existsSync(dumpPath)) return false
    const st = fs.statSync(dumpPath)
    if (st.size < 200) return false
  } catch {
    return false
  }
  return true
}

File: go-swarm/run_turn.go
package main

import (
	"fmt"
	"strings"
	"time"
)

func isContextSizeError(msg string) bool {
	m := strings.ToLower(msg)
	return strings.Contains(m, "bad request") ||
		strings.Contains(m, "context overflow") ||
		strings.Contains(m, "context length") ||
		strings.Contains(m, "too large") ||
		strings.Contains(m, "token") ||
		strings.Contains(m, "413") ||
		strings.Contains(m, "payload")
}

// isExternalAbortError: human TUI Esc / concurrent cancel — not stall.
func isExternalAbortError(msg string) bool {
	m := strings.ToLower(msg)
	if strings.Contains(m, "stall:") {
		return false
	}
	return strings.Contains(m, "abort") ||
		strings.Contains(m, "cancelled") ||
		strings.Contains(m, "canceled") ||
		strings.Contains(m, "interrupted")
}

// lastAssistantTextSDK reads last assistant text for a session (no AgentRecord).
func lastAssistantTextSDK(sdk *SDKClient, sessionID string) string {
	msgs, err := sdk.sessionMessages(sessionID, 0)
	if err != nil || len(msgs) == 0 {
		return ""
	}
	for i := len(msgs) - 1; i >= 0; i-- {
		msg := msgs[i]
		info, _ := msg["info"].(map[string]any)
		if info == nil {
			info = msg
		}
		role, _ := info["role"].(string)
		if role != "assistant" {
			continue
		}
		parts, _ := msg["parts"].([]any)
		var texts []string
		for _, p := range parts {
			pm, _ := p.(map[string]any)
			if pm == nil {
				continue
			}
			if t, _ := pm["type"].(string); t == "text" {
				if txt, _ := pm["text"].(string); txt != "" {
					texts = append(texts, txt)
				}
			}
		}
		return strings.Join(texts, "\n")
	}
	return ""
}

// turn runs a model turn with stall detection, external-abort soft recover, and rotate.
func (r *Run) turn(agent AgentRecord, prompt, systemPrompt string) (string, int, error) {
	var lastErr error
	// Local copy so rotate can update SessionID for retries.
	a := agent

	for attempt := 1; attempt <= maxTurnAttempts; attempt++ {
		t0 := time.Now()
		statuses, _ := r.sdk.sessionStatus()
		if s, ok := statuses[a.SessionID]; ok && s.Type != "idle" {
			r.waitForIdle(a.SessionID, 120)
		}

		pb := promptBody{
			Parts: []map[string]string{{"type": "text", "text": prompt}},
		}
		if systemPrompt != "" {
			pb.System = systemPrompt
		}
		pb.Model = &modelRef{ProviderID: ProviderID, ModelID: bareModel(a.Model)}

		err := r.sdk.sessionPromptAsync(a.SessionID, pb)
		if err != nil {
			lastErr = err
			r.log(fmt.Sprintf("  [host] turn error attempt %d/%d: %s", attempt, maxTurnAttempts, truncate(err.Error(), 300)))
			if r.handleTurnError(&a, err.Error(), attempt) {
				continue
			}
			return "", 0, lastErr
		}

		if stallErr := r.waitForIdleWithStall(a); stallErr != nil {
			lastErr = stallErr
			if stallErr == errStopped {
				return "", 0, errStopped
			}
			r.log(fmt.Sprintf("  [host] turn error attempt %d/%d: %s", attempt, maxTurnAttempts, truncate(stallErr.Error(), 300)))
			if r.handleTurnError(&a, stallErr.Error(), attempt) {
				continue
			}
			return "", 0, lastErr
		}

		if r.stopping.Load() || stopFileExists(r.paths.RunDir) {
			return "", 0, errStopped
		}

		r.watchMu.Lock()
		sw := r.systemWatch
		r.watchMu.Unlock()
		if sw != nil && sw.isWatchAbort() && a.Role == "worker" {
			return "", 0, fmt.Errorf("aborted: watch HOST: STOP (worker only)")
		}

		text := lastAssistantTextSDK(r.sdk, a.SessionID)
		secs := int(time.Since(t0).Seconds())
		oneLine := strings.TrimSpace(strings.Join(strings.Fields(text), " "))
		if oneLine != "" {
			r.log(fmt.Sprintf("  [reply:%s] %s", a.Role, truncate(oneLine, 300)))
		}
		r.log(fmt.Sprintf("  [metric] %s turn %ds", a.Role, secs))
		return text, secs, nil
	}
	if lastErr != nil {
		return "", 0, lastErr
	}
	return "", 0, fmt.Errorf("turn failed")
}

// handleTurnError returns true if caller should retry. May update agent.SessionID.
func (r *Run) handleTurnError(agent *AgentRecord, msg string, attempt int) bool {
	if r.stopping.Load() || strings.Contains(strings.ToLower(msg), "stopped") {
		return false
	}

	r.watchMu.Lock()
	sw := r.systemWatch
	r.watchMu.Unlock()
	if isExternalAbortError(msg) {
		if sw != nil && sw.isWatchAbort() {
			r.log(fmt.Sprintf("  [host] external abort on %s (watch/lead abort) — terminal, no re-prompt", agent.Role))
			return false
		}
		r.log(fmt.Sprintf("  [host] external abort on %s — wait idle, re-prompt same session (no rotate)", agent.Role))
		r.waitForIdle(agent.SessionID, 60)
		if attempt < maxTurnAttempts {
			time.Sleep(time.Duration(1500*attempt) * time.Millisecond)
			return true
		}
		return false
	}

	r.sdk.sessionAbort(agent.SessionID)
	r.waitForIdle(agent.SessionID, 60)

	if strings.Contains(strings.ToLower(msg), "stall:") {
		if attempt == 1 {
			r.log(fmt.Sprintf("  [host] stall soft-recover on %s — re-prompt same session (no rotate yet)", agent.Role))
			time.Sleep(2 * time.Second)
			return true
		}
		r.log("  [host] rotating session after repeated stall")
		r.rotateSession(agent)
		time.Sleep(1 * time.Second)
		return attempt < maxTurnAttempts
	}

	if isContextSizeError(msg) {
		r.log("  [host] rotating session after size/Bad Request")
		r.rotateSession(agent)
		time.Sleep(1 * time.Second)
		return attempt < maxTurnAttempts
	}

	if attempt == maxTurnAttempts-1 {
		r.rotateSession(agent)
	} else if attempt < maxTurnAttempts {
		time.Sleep(time.Duration(2*attempt) * time.Second)
	}
	return attempt < maxTurnAttempts
}

// rotateSession forks (preferred) or summarize→create fallback; updates agent.SessionID.
func (r *Run) rotateSession(agent *AgentRecord) {
	newID, err := r.sdk.sessionFork(agent.SessionID)
	if err == nil && newID != "" {
		r.persistSessionID(*agent, newID)
		agent.SessionID = newID
		r.log(fmt.Sprintf("  [host] session.fork ok for %s — new session %s", agent.Role, truncate(newID, 16)))
		return
	}
	r.log("  [host] session.fork failed: " + truncate(fmt.Sprintf("%v", err), 160))

	_ = r.sdk.sessionSummarize(agent.SessionID, modelRef{ProviderID: ProviderID, ModelID: bareModel(agent.Model)})
	summary := lastAssistantTextSDK(r.sdk, agent.SessionID)
	fresh, err := r.sdk.sessionCreate(fmt.Sprintf("swarm %s %s (rotated)", r.id, agent.Role))
	if err != nil {
		r.log("  [host] rotate create failed: " + truncate(err.Error(), 160))
		return
	}
	r.persistSessionID(*agent, fresh)
	agent.SessionID = fresh
	note := "[host] Prior session was rotated. Read MISSION.md, DIALOGUE.md, MEMORY.md for continuity."
	if strings.TrimSpace(summary) != "" {
		note = "[host] Prior session rotated. Continuity summary:\n\n" + truncate(summary, 6000)
	}
	r.sdk.sessionInjectContext(fresh, note, &modelRef{ProviderID: ProviderID, ModelID: bareModel(agent.Model)})
	r.log(fmt.Sprintf("  [host] rotated session for %s (fallback path)", agent.Role))
}

func (r *Run) persistSessionID(agent AgentRecord, newID string) {
	r.recMu.Lock()
	defer r.recMu.Unlock()
	if r.rec != nil {
		for i := range r.rec.Agents {
			if r.rec.Agents[i].SessionID == agent.SessionID ||
				(r.rec.Agents[i].Role == agent.Role && r.rec.Agents[i].Name == agent.Name) {
				r.rec.Agents[i].SessionID = newID
			}
		}
		regSave(r.rec)
	}
	for i := range r.workers {
		if r.workers[i].SessionID == agent.SessionID || r.workers[i].Name == agent.Name {
			r.workers[i].SessionID = newID
		}
	}
}

// waitForIdle polls session status until idle or timeout.
func (r *Run) waitForIdle(sessionID string, timeoutSecs int) {
	deadline := time.Now().Add(time.Duration(timeoutSecs) * time.Second)
	seenBusy := false
	started := time.Now()
	for time.Now().Before(deadline) {
		if r.stopping.Load() || stopFileExists(r.paths.RunDir) {
			return
		}
		statuses, err := r.sdk.sessionStatus()
		if err != nil {
			time.Sleep(2 * time.Second)
			continue
		}
		st, ok := statuses[sessionID]
		if !ok || st.Type == "idle" {
			if seenBusy {
				return
			}
			// Never became busy: give up after 90s
			if time.Since(started) > 90*time.Second {
				return
			}
			time.Sleep(2 * time.Second)
			continue
		}
		if st.Type == "busy" || st.Type == "retry" || st.Type == "working" {
			seenBusy = true
		}
		time.Sleep(2 * time.Second)
	}
}

// waitForIdleWithStall waits for idle; returns stall error if bus quiet too long.
// Emits periodic alive heartbeats so long model/tool turns do not look dead in the terminal.
func (r *Run) waitForIdleWithStall(agent AgentRecord) error {
	deadline := time.Now().Add(12 * time.Hour)
	seenBusy := false
	started := time.Now()
	sessionID := agent.SessionID
	lastLogStallHint := time.Time{}
	lastAliveLog := time.Time{}
	stallMs := int64(stallThreshold / time.Millisecond)
	const aliveEvery = 45 * time.Second

	for time.Now().Before(deadline) {
		if r.stopping.Load() || stopFileExists(r.paths.RunDir) {
			return errStopped
		}
		r.watchMu.Lock()
		sw := r.systemWatch
		r.watchMu.Unlock()
		if sw != nil && sw.isWatchAbort() && agent.Role == "worker" {
			return fmt.Errorf("aborted: watch HOST: STOP (worker only)")
		}

		// Session id may have been rotated mid-flight elsewhere — use current
		sessionID = agent.SessionID

		statuses, err := r.sdk.sessionStatus()
		if err != nil {
			if time.Since(lastAliveLog) > aliveEvery {
				r.log(fmt.Sprintf("  [host] %s waiting — status poll error (will retry): %s", agent.Role, truncate(err.Error(), 120)))
				lastAliveLog = time.Now()
			}
			time.Sleep(2 * time.Second)
			continue
		}
		st, ok := statuses[sessionID]
		if !ok || st.Type == "idle" {
			if seenBusy {
				return nil
			}
			if time.Since(started) > 90*time.Second {
				return nil
			}
			if time.Since(lastAliveLog) > aliveEvery {
				r.log(fmt.Sprintf("  [host] %s waiting to become busy… elapsed %ds", agent.Role, int(time.Since(started).Seconds())))
				lastAliveLog = time.Now()
			}
			time.Sleep(2 * time.Second)
			continue
		}
		if st.Type == "busy" || st.Type == "retry" || st.Type == "working" {
			seenBusy = true
		}

		if r.bus != nil && seenBusy {
			lastBus := r.bus.lastActivityFor(sessionID)
			quietMs := int64(0)
			if lastBus > 0 {
				quietMs = time.Now().UnixMilli() - lastBus
			} else {
				// No bus events since start of wait — use time since seen busy
				quietMs = time.Since(started).Milliseconds()
			}

			half := stallMs / 2
			if half > int64(staleRunningToolQuiet/time.Millisecond) {
				half = int64(staleRunningToolQuiet / time.Millisecond)
			}
			if r.bus.clearStaleRunningTools(sessionID, half) {
				r.log(fmt.Sprintf("  [host] cleared stale running-tool flag on %s (no bus events)", agent.Role))
			}

			// Periodic "still alive" so Ctrl+C isn't the default when thinking is quiet.
			if time.Since(lastAliveLog) > aliveEvery {
				tools := r.bus.runningToolCount(sessionID)
				r.log(fmt.Sprintf("  [host] %s still working — status=%s bus_quiet=%ds tools_running=%d elapsed=%ds (auto-stall at %dm; do not Ctrl+C)",
					agent.Role, st.Type, quietMs/1000, tools, int(time.Since(started).Seconds()), stallMs/60_000))
				lastAliveLog = time.Now()
			}

			if r.bus.hasRunningTools(sessionID) && quietMs < int64(2*time.Minute/time.Millisecond) {
				time.Sleep(5 * time.Second)
				continue
			}

			if quietMs >= stallMs {
				detail := "no bus events"
				if st.Type != "" {
					detail = "status still " + st.Type
				}
				return fmt.Errorf("stall: no OpenCode bus events for %dm on %s (%s; threshold %dm)",
					quietMs/60_000, agent.Role, detail, stallMs/60_000)
			}

			if quietMs >= int64(5*time.Minute/time.Millisecond) && time.Since(lastLogStallHint) > 5*time.Minute {
				r.log(fmt.Sprintf("  [host] %s quiet ~%dm (stall at %dm) — host will soft-recover; leave it running", agent.Role, quietMs/60_000, stallMs/60_000))
				lastLogStallHint = time.Now()
			}
		} else if seenBusy && time.Since(lastAliveLog) > aliveEvery {
			r.log(fmt.Sprintf("  [host] %s still working — status=%s elapsed=%ds", agent.Role, st.Type, int(time.Since(started).Seconds())))
			lastAliveLog = time.Now()
		}

		time.Sleep(5 * time.Second)
	}
	return fmt.Errorf("stall: waitForIdle exceeded wall clock on %s", agent.Role)
}

func (r *Run) lastAssistantText(agent AgentRecord) string {
	return lastAssistantTextSDK(r.sdk, agent.SessionID)
}
